### PR TITLE
feat: Implement Renegade liquidity module

### PR DESCRIPTION
**Pull request by joeykraut**
## **Protocol Information**

- **Protocol Name:** Renegade
- **Protocol Website:** [website](https://renegade.fi/), [docs](https://docs.renegade.fi/)
- **Indexing Smart Contract Addresses (e.g. factories):** n/a

---

## **Summary of Integration**

Provide a brief overview of your integration:

- **What does your protocol do?**
    
    Renegade is the first on-chain, decentralized [darkpool](https://en.wikipedia.org/wiki/Dark_pool). We run a decentralized, private order book [off-chain](https://docs.renegade.fi/core-concepts/mpc-zkp#network-architecture) (guaranteeing pre-trade privacy), and settle on-chain via [zero-knowledge powered](https://docs.renegade.fi/core-concepts/mpc-zkp#the-commit-reveal-scheme) smart contracts (guaranteeing post-trade privacy).
    
    Renegade's [external match](https://docs.renegade.fi/technical-reference/typescript-sdk#external-atomic-matching) feature allows users to opt-in to crossing their darkpool orders against external orders -- effectively foregoing post-trade privacy to tap into broader liquidity. This feature allows routers, solvers, etc (in this case GlueX) to tap into dark liquidity.
    
- **What type of liquidity does it provide (DEX, lending, yield farming, etc.)?**
    
    Spot DEX liquidity, pegged to the real-time Binance midpoint price.
    
- **Any unique features that GlueX should consider?**
    - Prices in our quotes are firm. There is no notion of slippage or price impact in Renegade; rather, quotes always execute atomically at the advertised price. For this reason, the Renegade external match API can feel a bit like an RFQ or MM api, despite that Renegade is fundamentally a DEX.
    - Renegade is a [crossing network](https://en.wikipedia.org/wiki/Crossing_network); so there is no maker-taker dynamic or bonding curve to consider in the integration. We cross natural counterflow at the Binance midpoint.
    - Because the orderbook is private, there is no price discovery in Renegade, so we peg our prices to the Binance midpoint. You can be sure that a quote received from Renegade is pegged to the realtime midpoint price.

---

### **Execution Functions Required**

The Renegade API will route orders through one of two contracts:

- **The [gas sponsorship](https://docs.renegade.fi/technical-reference/typescript-sdk#gas-sponsorship) contract**: If eligible, the order will be routed through the gas sponsorship contract. This contract will execute the swap then refund the gas cost of the swap to the configured address in the output token.
- **The darkpool contract**: If an order is not eligible for gas sponsorship, it will be routed directly to the Renegade darkpool contract. This contract execute the swap directly without gas rebate.

For the gas sponsorship contract, the relevant methods are:

1. `sponsorAtomicMatchSettle`: executes the swap and refunds the gas cost. The receiver of the output token is assumed to be `msg.sender`. Returns the amount of the output token sent to the receiver.

```
pub fn sponsor_atomic_match_settle_with_refund_options(
    &mut self,
    receiver: Address,
    internal_party_match_payload: Bytes,
    valid_match_settle_atomic_statement: Bytes,
    match_proofs: Bytes,
    match_linking_proofs: Bytes,
    refund_address: Address,
    nonce: U256,
    refund_native_eth: bool,
    refund_amount: U256,
    signature: Bytes,
)
    external
    payable
    returns (uint256)
```

You can find the [code here](https://github.com/renegade-fi/renegade-contracts/blob/1561abe562375238acc06b8fec2b2bdf765ca3f4/contracts-stylus/src/contracts/gas_sponsor.rs#L313).

For the darkpool contract, the relevant methods are:

1. `processAtomicMatchSettle`: Processes an external match and handles ERC20 transfers for the swap. Returns the amount of the output token that was swapped.

```
function processAtomicMatchSettle(
    bytes memory internal_party_match_payload,
    bytes memory valid_match_settle_atomic_statement,
    bytes memory match_proofs,
    bytes memory match_linking_proofs
) external payable returns (uint256);
```

With [code here](https://github.com/renegade-fi/renegade-contracts/blob/1561abe562375238acc06b8fec2b2bdf765ca3f4/contracts-stylus/src/contracts/darkpool.rs#L576).

1. `processAtomicMatchSettleWithReceiver`: Processes an external match and handles ERC20 transfers for the swap. This method allows the caller to specify a receiver of the output tokens that is not `msg.sender`. Returns the number of output tokens sent to the receiver.

```
function processAtomicMatchSettleWithReceiver(
    address receiver,
    bytes memory internal_party_match_payload,
    bytes memory valid_match_settle_atomic_statement,
    bytes memory match_proofs, bytes memory match_linking_proofs
) external payable returns (uint256);
```

With [code here](https://github.com/renegade-fi/renegade-contracts/blob/1561abe562375238acc06b8fec2b2bdf765ca3f4/contracts-stylus/src/contracts/darkpool.rs#L603).

### **Functions Implemented**

Following the template laid out in [#8](https://github.com/gluexprotocol/liquidity-module-self-integration/pull/8), we've implemented:

- `get_sell_quote`: We convert the order into a USDC-quoted midpoint pegged order and fetch an `AtomicMatchApiBundle` directly from the API, which contains
    - `send` - the amount sent by the user, net of fees and gas sponsorship
    - `receive` - the amount received by the user, net of fees and gas sponsorship
    - `settlement_tx` - the transaction to be submitted on-chain
        - `data` - calldata
        - `to` - darkpool contract address / gas sponsorship contract address if the quote is sponsored (enabled by default)
- `get_buy_quote`: Similar to `get_sell_quote` we convert the order into a USDC-quoted midpoint pegged order and fetch directly from the API. For buy side quotes, we specify either an `exact_base_amount` or an `exact_quote_amount` based on the direction of the quoted pair. Return type is the exact same shape.

**Note**: Currently all pairs in Renegade are USDC quoted, hence the pair conversion logic.

### **Dynamic States Required for AMM Calculations**

n/a

### **Static States Required for AMM Calculations**

n/a

### **Dependencies**

- `httpx` - Async HTTP client for making authenticated requests to Renegade APIs
- `pydantic` - Data validation and serialization models for API request/response structures
- `web3` - Ethereum address checksumming and blockchain transaction type definitions
- `typing-extensions` - type safety across Python versions

### **Other Requirements**

- **API Keys:** I will send an API key via Telegram to the team in our group chat.
- **Special Access Requirements:** No "special" requirements, other than the standard ERC20 approvals for all send tokens.

---

## **Test Results**

Provide details of tests performed, including:

- **Test logs (paste relevant results):**

image

- **Edge cases considered**: The integration is very thin, only requiring pair conversion. I added test cases for each pair conversion case: buy side output amount, sell side output amount, buy side input amount, and sell side input amount.
- **Pass/fail outcomes**: All tests pass.

### CODE REVIEW

File: modules/renegade_liquidity_module.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    def _get_client(self, fixed_parameters: Dict) -> ExternalMatchClient:
+        """
+        Gets the Renegade client for the given fixed parameters.
+        """
+        chain = fixed_parameters.get("chain")
+        api_key = fixed_parameters.get("api_key")
+        api_secret = fixed_parameters.get("api_secret")
+
+        if chain == Chain.ARBITRUM_ONE:
```
Comment (sarveshotex) [inline modules/renegade_liquidity_module.py:135]: I see you are using an external Sdk for creating the client. Is it possible to have a simple API endpoint instead. Its not ideal to have an SDK installed for every protocol for us.

File: modules/renegade_liquidity_module.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        return True
+
+    def _check_usdc(self, token: Token) -> bool:
+        """
+        Checks if the token is USDC.
+
+        :param token: The token to check.
+        :return: True if the token is USDC, False otherwise.
+        """
+        return token.symbol.upper() == "USDC"
```
Comment (sarveshotex) [inline modules/renegade_liquidity_module.py]: symbol is not a required key for Token class internally and symbol is not a unique feature for tokens. We would suggest not using it for any validation.

File: modules/renegade_liquidity_module.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        # If the quote is not found, return None
+        if not signed_quote:
+            return None, None, None
+        
+        # Fees are taken out of the receive amount, so the input token fee is zero
+        # The output token amount on the quote accounts for this fee
+        return NO_INPUT_FEE, signed_quote.quote.send.amount, signed_quote
+
+    async def assemble_quote(
```
Comment (sarveshotex) [inline modules/renegade_liquidity_module.py]: This doesn't seemed to be used.

**Review by sarveshotex — CHANGES_REQUESTED** [review]

File: modules/renegade_liquidity_module.py (hunk 1 above)
Comment (sehyunc) [reply modules/renegade_liquidity_module.py:135]: Done! We no longer depend on `renegade-sdk` and instead hit the API directly.

File: modules/renegade_liquidity_module.py (hunk 2 above)
Comment (sehyunc) [reply modules/renegade_liquidity_module.py]: Sounds good, our validation now checks against a list of known USDC addresses.

File: modules/renegade_liquidity_module.py (hunk 3 above)
Comment (sehyunc) [reply modules/renegade_liquidity_module.py]: Removed!
